### ISSUE TITLE
Unable to preventDefault inside passive event listener invocation.

### ISSUE THREAD

**Issue by glushkova91** [focus-trap/focus-trap#1165]
Hello,

We have an use case of using this library together with angular cdk a11y FocusMonitor module, and in this case the first focusable element is unreachable by `Tab` key.

As I understood the reason is that the focus-trap library uses `preventDefault` inside keydown event handler in order to not let browser natively navigate to the next focusable element after programmatically setup the focus on the first element:
https://github.com/focus-trap/focus-trap/blob/master/index.js#L762

In the same time `InputModalityDetector` class, which is used by `FocusMonitor`, adds event listener to keydown event with configuration `passive: true`
https://github.com/angular/components/blob/main/src/cdk/a11y/input-modality/input-modality-detector.ts#L196
which prevents `preventDefault`.

I'm not sure if this problem can and would be considered for fix, since obviously neither library can be integrated with everything, and sometimes the most logical solution is to remove one of the libraries, but in our case we need both of them.
I didn't come up with any better suggestion as consider tracking focus event instead of keydown event (so to let browser focus on the wrong element first and then return focus back to trapped elements)

I would be happy to hear any feedback.

**Comment by stefcameron** [comment]
@glushkova91 Thanks for reaching out! I'll say that I'm not at all familiar with this `InputModalityDetector`, nor with `FocusMonitor`. What exactly is their purpose? I get a bit suspicious when I see multiple focus management libraries used together because they tend to compete and things tend to not go very well.

Focus-trap creates a non-passive `keydown` event handler on the `document` while you say that `InputModalityDetector` creates a passive one (I can't see that in the code; it's buried somewhere in the options and isn't immediately apparent to me).

The two event listeners are completely separate, so focus-trap is free to `preventDefault()` while `InputModalityDetector` is not (and doesn't anyway).

So far, so good, it would seem.

But now you're thinking that focus-trap is preventing the first tabbable (and so focusable) node in its trap from getting focus because it's calling `preventDefault()`?

This I don't understand. The focus-trap code you pointed to aims to prevent default only when there is a `destinationNode`, and that would only be when pressing TAB from one of the edge nodes in a focus-trap container. An edge node is either the last node + TAB (moving forward), or the first node + SHIFT+TAB (moving backward), and the result would be setting focus to the first or last node, respectively.

Preventing default is necessary in this case because the browser will want to move focus out of the container, and thus out of the trap, and if we let it escape and then pull it back in, the wrong node will temporarily get focus and that could cause things like unwanted page scrolling (if the node that temporarily gets focus happens to be scrolled out of view). It could also trigger a `focus` event handler you don't expect to get triggered while the trap is active, executing code you don't want to run while a dialog is displayed or whatever you're using the trap for.

Also note that when a trap is activated, it seeks to set focus on the first node in the trap. This is another point of confusion with the behavior you're describing, stating that the first node is unreachable with the TAB key.

This is my understanding so far. Let me know what you think!

**Comment by glushkova91** [comment]
Hello @stefcameron.
TLDR I believe it's Zone.js issue, not focus-trap issue. More details below:

First of all I prepared a visual demonstration of this problem
https://stackblitz.com/edit/stackblitz-starters-mvryhu
(Angular app starter, added an empty event listener with `passive: true` config, and as you can see the first link is not accessible via Tab)

Then get back to you comment:
1. FocusMonitor is a service from CDK to subscribe to focus changes and get some details about focus status, but in fact it's not that important, we can simply add 
    document.addEventListener('keydown', () => {}, {
      passive: true,
      capture: true,
    });
    before `focusTrap` activation in order to reproduce the problem.
2. You're right that you're suspicious about multiple libraries with similar purposes, in our case it's indeed the reason of the problem. But we can't just get rid of Angular CDK because it's a powerful library AND in the same time we need a focus-trap library for UI common components package which is injected into Angular app. We need to keep UI components independent from any framework because it's injected also into Vue and React apps.
3. Link to passive event config in `InputModalityDetector` https://github.com/angular/components/blob/main/src/cdk/a11y/input-modality/input-modality-detector.ts#L73
4. You are also right that 2 event listeners with their own configs should be completely separate. BUT as I was able to understood here zone.js comes to a game. It collects and store all event listeners in its state and creates a single event listener for browser with options of the first event (gonna be `passive: true` in our case) and when keydown event happen it substitute non-passive keydown event config from focus-trap with passive event config.
5. And if the event listener `checkKey` from focus trap would be called with `passive: true` config then `preventDefault` is gonna fail

**Comment by glushkova91** [comment]
There is an [open ticket](https://github.com/angular/angular/issues/45020) for that in Angular repo and [draft PR](https://github.com/angular/angular/pull/49477). I'm closing the ticket, feel free to update "Known issues" though.

**Comment by stefcameron** [comment]
@glushkova91 I really appreciate your sandbox repro. I can indeed see the first link gets skipped in both TAB directions. I also appreciate your investigation into `zone.js` I believe you have found the root cause. Hopefully that PR gets some attention and is eventually merged.

Is `zone.js` included as part of the Angular CDK you need to use? If not, what's the purpose of using it? I actually remember being _at_ ng-conf 2014 in Salt Lake City (UT) when this new zone concept was revealed but AFAICT it never took off, except perhaps within Angular circles.

This will help me formulate something to write about in the README for [known issues](https://github.com/focus-trap/focus-trap?tab=readme-ov-file#help).

**Comment by stefcameron** [comment]
It was still bugging me why that first node was being excluded and now I understand because of this (looking at the source in your sandbox):

<img width="611" alt="image" src="https://github.com/focus-trap/focus-trap/assets/2855350/cb16a43d-33e5-46cc-ba2a-5f98be8ca89d">

To be clear, an exception is not thrown when focus-trap calls `event.preventDefault()`. The browser simply posts an error-level console message, which DevTools then flags in the source in a way that looks similar to how it flags the source of an exception.

So because of the interference of `zone.js`, focus-trap's call to `event.preventDefault()` is ignored. focus-trap then proceeds to call `.focus()` on "Link 1" -- as it should, when you're pressing TAB _from_ "Link 3" which is an edge node, as I've described in my previous comment, requiring the focus to "wrap" to the start of the container -- but this too is effectively ignored because the browser is simply waiting for the event listener's code to finish executing and then promptly sets focus to the next node in tab order.

And since focus-trap has gone and focused "Link 1", the next node is "Link 2", making it look to the user like "Link 1" can't be focused.

But if you use SHIFT+TAB instead, to go in reverse tab order, then it's "Link 3" that can't be focused while "Link 1" can now be focused.

And by saying "can't be focused", what's really happening is that "Link 1" or "Link 3" (depending on tab direction) __does__ get focused, but the browser's default behavior isn't prevented (because of `zone.js` interference) and focus is moved _again_ to the next node.

Another way to put it is that when you're on an edge node and you press TAB, you'll see not one, but _two_ `focus` events: The first one will be on the first node in tab order in the container (because of focus-trap's code) and the second/final one will be on the second node in tab order (because of the browser).

I'm not sure what, if anything, focus-trap could reasonably do (with good results) in order to mitigate this given focus-trap has good reason to be interfering with normal browser focus behavior on edge nodes (as I've previously explained).

**Comment by glushkova91** [comment]
@stefcameron sorry for the late response. Answering your question about zone.js:

Zone.js is a library which is a part of Angular bundle, not Angular CDK. It's implemented by angular team for angular framework, though it can be installed and used separately https://www.npmjs.com/package/zone.js. It helps to detect that application state has been changed. Mostly it's a proxy to asynchronous events (timeout, browser events, http calls etc), zone.js captures them and angular updates application state based on it.

Regarding the last message: yes, you described the problem very precisely. 
Regarding "if anything can be done from focus-trap side": 
First of all, after research I believe there is no benefit of using `passive: true` for `keydown` event as it's done in Angular CDK. I couldn't find any recommendation of using it for `keydown` and also found some performance testing article https://gist.github.com/iamakulov/45803e89db2eb44a7a6be33a80ffcab7. So the approach you use in focus-trap has no observed downside for performance or any other reasons to change it from my point of view. 

Second thing, one of the possible solution to make the `focus-trap` compatible with angular a11y cdk and zone.js might be introducing start and end anchor in the beginning of focus trap area and catch focus on them. Saw this approach in angular a11 cdk https://github.com/angular/components/blob/main/src/cdk/a11y/focus-trap/focus-trap.ts#L96. But it will require the major changes in your library, not sure if it worths it especially in the case there is no issue on your side.

**Comment by stefcameron** [comment]
@glushkova91 Thank you for confirming, and for your thoughts/suggestions. For now, adding anchors to Focus-trap feels like too much of a lift for resulting benefits. I did, however, [document the issue](https://github.com/focus-trap/focus-trap/blob/master/README.md#first-element-in-trap-is-unreachable-with-the-tab-key) in the Help section in case others face this issue as well.

**Comment by stefcameron** [comment]
@all-contributors add @glushkova91 for docs
